Add ReasonML layer

actionshrimp opened the pull request:
This adds a layer for [`reason`, aka `reasonml`](https://reasonml.github.io/), based primarily on `reason-mode`, but also on `merlin` via the `ocaml` layer.

I didn't actually write the original version, this is based on @fredyr's version mentioned in the official language discussion here - https://github.com/facebook/reason/issues/1149 - that discussion seems to have stalled a bit as I think the majority of people currently use vscode - as far as I can tell there hasn't been any other spacemacs-project specific discussion so far, please let me know if I've missed it.

I've been using that layer for a few months now and it seems to work well, based on `reason-mode`, so I took the liberty of bringing it up to date slightly and add a few docs around the most commonly used features, along with some setup instructions.

varunpatro commented:
This is really useful, thanks!

actionshrimp commented:
Just added a couple of new things:
- jumping between JSX tags with `%` using evil-matchit, and some refmt keybindings to convert a region
- a couple of new refmt keybindings for converting re->ml and ml->re

duianto commented on layers/+lang/reason/README.org:
```diff
@@ -0,0 +1,70 @@
+
+** Features:
+  - Syntax highlighting / indentation
+  - Autocomplete (via `merlin`)
```
The markdown inline code (surrounding backquotes) syntax needs to be changed to org verbatim (surrounding equal signs), throughout the readme.org file.
#### before:
`` `merlin` ``
#### after:
`=merlin=`

duianto commented on layers/+lang/reason/README.org:
```diff
@@ -0,0 +1,70 @@
+You'll need merlin (binaries `ocamlmerlin` and `ocamlmerlin-reason`) installed
+on your system and on your `PATH` for errors and autocompletion.
+
+If you're familiar with `opam` you can `opam install merlin reason` to get
```
Commands to run are easier to select/copy if they are wrapped in a source block:
```sh
#+BEGIN_SRC sh
  opam install merlin reason
#+END_SRC
```

duianto commented on layers/+lang/reason/README.org:
```diff
@@ -0,0 +1,70 @@
+If your project depends on a specific version of `refmt`, you can set
+`refmt-command` (via `customize-mode` or `(setq refmt-command ..)`) to the path
+to a particular binary. You can also use the special values `'npm` or `'opam` to
+run `refmt` via `npx` or `opam exec`, which will keep you on the correct version
```
typo: you -> your
#### before:
for you current opam switch

#### after:
for your current opam switch

duianto commented on layers/+lang/reason/README.org:
```diff
@@ -0,0 +1,70 @@
+| ~SPC m = =~ | Refmt |
+| ~SPC m t r~ | Toggle auto-refmt on save |
+
+# Use GitHub URLs if you wish to link a Spacemacs documentation file or its heading.
```
Removed the template examples.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+;;; License: GPLv3
+
+(defconst reason-packages
+  '(company
```
Sort packages alphabetically.

And the functions should probably be sorted alphabetically based on the package name.

Possibly like this:
#### before:
(defun reason/post-init-company ()
(defun reason/init-reason-mode ()
(defun reason/post-init-evil-matchit ()
(defun reason/pre-init-popwin ()
(defun reason/pre-init-utop ()
(defun reason/post-init-merlin ()
(defun flycheck-ocaml-reason-setup ()
(defun reason/post-init-flycheck ()
(defun reason/post-init-flycheck-ocaml ()
(defun ocaml/init-flycheck-ocaml ()


#### after:
(defun reason/post-init-company ()
(defun reason/post-init-evil-matchit ()
(defun reason/post-init-flycheck ()
(defun reason/post-init-flycheck-ocaml ()
(defun ocaml/init-flycheck-ocaml ()
(defun flycheck-ocaml-reason-setup ()
(defun reason/post-init-merlin ()
(defun reason/pre-init-popwin ()
(defun reason/init-reason-mode ()
(defun reason/pre-init-utop ()

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+    (spacemacs|add-company-backends
+      :backends merlin-company-backend
+      :modes reason-mode))
+  )
```
Join trailing parenthesis.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+    :mode ("\\.rei?\\'" . reason-mode)
+    :init
+    (progn
+      (add-hook 'reason-mode-hook (lambda ()
```
sdwolfz said about this hook within a hook:
>Reformating on save should be controlled by a variable, disabled by default

actionshrimp replied:
I've had a go at restructuring this a bit and turned it off by default, with a note in the README about how to turn it on with a layer variable.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+        :status reason-auto-refmt
+        :on (setq reason-auto-refmt t)
+        :off (setq reason-auto-refmt nil))
+      )
```
Join trailing parenthesis.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+       (setq utop-prompt 'reason/rtop-prompt)
+       (setq utop-initial-command "let myVar = \"Hello Reason!\";")
+       (setq utop-phrase-terminator ";")
+       ))
```
Join trailing parentheses.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+      (spacemacs/set-leader-keys-for-major-mode 'reason-mode
+        "er" 'utop-eval-region
+        "eb" 'utop-eval-buffer
+        "ee" 'utop-eval-phrase)
```
Join trailing the parentheses on the two lines below.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+        "hh" 'merlin-document
+        "ht" 'merlin-type-enclosing
+        "hT" 'merlin-type-expr
+        "rd" 'merlin-destruct)
```
Join trailing parentheses.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+(defun flycheck-ocaml-reason-setup ()
+  (with-eval-after-load 'merlin
+    (setq merlin-error-after-save nil)
+
```
Remove empty line.
```suggestion
```

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+
+    (flycheck-define-generic-checker 'reason-merlin
+      "A syntax checker for Reason using Merlin Mode.
+
```
Remove empty line.
```suggestion
```

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+      :start #'flycheck-ocaml-merlin-start
+      :verify #'flycheck-verify-ocaml-merlin
+      :modes '(reason-mode)
+      :predicate (lambda () (and merlin-mode
```
The `and` statement probably doesn't need a lambda wrapped around it here.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+                                 ;; Don't check if Merlin's own checking is
+                                 ;; enabled, to avoid duplicate overlays
+                                 (not merlin-error-after-save))))
+
```
Remove empty line.
```suggestion
```

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+    (interactive)
+    (add-to-list 'flycheck-checkers 'reason-merlin)))
+
+(when (configuration-layer/layer-used-p 'syntax-checking)
```
#### First:
Each package that's defined for a layer, probably needs to have an init function defined. Therefore the following functions can't be wrapped in this when statement.

That might be why I get these warning messages on startup:
>Warnings:
    - package flycheck not initialized in layer reason, you may consider removing this
      package from the package list or use the :toggle keyword instead of a \`when'
      form.
    - package flycheck-ocaml not initialized in layer reason, you may consider
      removing this package from the package list or use the :toggle keyword instead
      of a \`when' form.
    - Cannot find "opam" or "merlin" executable. The ocaml layer won't work properly.

The last line is most likely unrelated, I just haven't installed `opam` or `merlin`.

#### Second:
There's something strange going on with this line.
```elisp
(when (configuration-layer/layer-used-p 'syntax-checking)
```

It returns `nil` even though `syntax-checking` is listed in the `dotspacemacs-configuration-layers` variable.

This can be seen by adding these lines above the when statement:
```elisp
(message "dotspacemacs-configuration-layers: %s"
         dotspacemacs-configuration-layers)
(message "(configuration-layer/layer-used-p 'ocaml): %s"
         (configuration-layer/layer-used-p 'ocaml))
(message "(configuration-layer/layer-used-p 'syntax-checking): %s"
         (configuration-layer/layer-used-p 'syntax-checking))
```

When Spacemacs is restarted, then the messages buffer shows:
```
dotspacemacs-configuration-layers: (autohotkey colors emacs-lisp git helm markdown multiple-cursors ocaml org reason spell-checking syntax-checking treemacs version-control)
(configuration-layer/layer-used-p ’ocaml): (ocaml multiple-cursors markdown helm git emacs-lisp colors autohotkey)
(configuration-layer/layer-used-p ’syntax-checking): nil
```

`ocaml` is found but not `syntax-checking`

---



Each init function that requires the `syntax-checking` layer might need to check if it exists individually.

Same thing with `(configuration-layer/layer-used-p 'ocaml)` below.

#### System Info :computer:
- OS: windows-nt
- Emacs: 26.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 642eaf11f), **This PR has been applied on: b80154bd52b83b0877c6de24b0a73b48661f5ec5**
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey colors emacs-lisp git helm markdown multiple-cursors ocaml org reason spell-checking syntax-checking treemacs version-control)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS LCMS2

actionshrimp replied:
For your second point: I think I've run into something similar before - the order of the layers matters, so syntax-checking has to come before reason in your layer list for this to work.

I'll have a play around with this syntax-checking check anyway, I borrowed it from the ocaml layer but it looks like other layers dont do this the same way, will see if I can simplify.

duianto commented on layers/+lang/reason/packages.el:
```diff
@@ -0,0 +1,160 @@
+(defun reason/pre-init-popwin ()
+  (spacemacs|use-package-add-hook popwin
+    :post-config
+    (push '("*Refmt Errors*" :tail t :position bottom :noselect t) popwin:special-display-config)))
```
Keep this line below 80 chars by adding a newline.

#### Before
```elisp
    (push '("*Refmt Errors*" :tail t :position bottom :noselect t) popwin:special-display-config)))
```
#### After
```elisp
    (push '("*Refmt Errors*" :tail t :position bottom :noselect t)
          popwin:special-display-config)))
```

duianto commented on layers/+lang/reason/README.org:
```diff
@@ -0,0 +1,70 @@
+
+The main keybindings, see `packages.el` for the main list.
+
+| Key Binding | Description    |
```
Pressing `TAB` with the cursor in the key binding table, in a `org-mode` buffer formats the table to:
```org
| Key binding | Description                       |
|-------------+-----------------------------------|
| ~SPC m g g~ | Jump to definition                |
| ~SPC m g G~ | Jump to definition (other window) |
| ~SPC m g b~ | Jump back                         |
| ~SPC m h t~ | Show type                         |
| ~SPC m h h~ | Show docs                         |
| ~SPC m r d~ | Merlin destruct                   |
| ~SPC m = =~ | Refmt                             |
| ~SPC m t r~ | Toggle auto-refmt on save         |
```

And `Key Binding` is written with a lower case `b`: `Key binding`
The documentation bot will correct it, but we might as well fix it while making other changes.

actionshrimp commented:
Thanks very much for taking the time to go through this! Will get this tidied up ASAP with your suggestions.

actionshrimp commented:
Think I've incorporated everything you suggested! Have also renamed the layer to `reasonml` to make it a bit clearer what it's for for anyone unfamiliar with the ecosystem (it's a lot easier to google!).

duianto commented:
The `config.el`, `funcs.el` and `layers.el` files, probably need a header. All of the 5+ layer files I just checked seems to have one.

The `html` layer uses all three files so they can be used as examples:
https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Blang/html/config.el
https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Blang/html/funcs.el
https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Blang/html/layers.el

It seems like it's just the first line that changes.

duianto commented:
It would probably useful to also add a new layer `ReasonML` section to:
https://github.com/syl20bnr/spacemacs/blob/develop/CHANGELOG.develop#layer-changes-and-fixes
with the key bindings and their descriptions.

Possibly something like this between the `React` and `Ruby` layer entries:
```org
**** ReasonML
- Key bindings (thanks to fredyr and Dave Aitken):
  - ~SPC m g g~ Jump to definition
  - ~SPC m g G~ Jump to definition (other window)
  - ~SPC m g b~ Jump back
  - ~SPC m h t~ Show type
  - ~SPC m h h~ Show docs
  - ~SPC m r d~ Merlin destruct
  - ~SPC m = =~ Refmt
  - ~SPC m t r~ Toggle auto-refmt on save
```

Then someone that's reading through the changelog will be able to see what some of the layers capabilities are.

actionshrimp commented:
All done, thanks again.

duianto commented on CHANGELOG.develop:
```diff
@@ -2224,6 +2225,20 @@ Other:
   - Use JSX header without breaking React (thanks to Jam Risser)
   - Fixed React imports when using web beautify. (thanks to Ismael)
   - Made layer depends on node and tern layers (thanks to Sylvain Benner)
+
```
Match the other layer entries by removing the empty lines before and after the ReasonML section.

duianto commented:
Squash the commits into one. We try to keep the commits to one per PR (unless they are complex changes).

That should be the last suggestion, then this should be ready to go.

actionshrimp commented:
👍

duianto commented:
Thank you for contributing to Spacemacs.
The PR has been cherry-picked into the `develop` branch. It's safe to delete your branch.
Commit: https://github.com/syl20bnr/spacemacs/commit/f40361fd64950f9eece1be792d02369d4e15552d
